### PR TITLE
Migrate pallet-randomness-collective-flip to pallet attribute macro

### PR DESCRIPTION
**Pull request by koushiro**
Part of #7882 

Migrate the `randomness-collective-flip` pallet to the new pallet attribute macro

polkadot companion: paritytech/polkadot#3200

## Breaking Changes

need to add code like `impl pallet_randomness_collective_flip::Config for Runtime {}`.

### CODE REVIEW

File: frame/randomness-collective-flip/src/lib.rs [hunk 1]
```diff
@@ -86,20 +109,19 @@ decl_module! {
 			});
 
 			0
```
Comment (shawntabrizi) [inline frame/randomness-collective-flip/src/lib.rs:111]:
```suggestion
			T::DbWeight::get().reads_writes(1, 1)
```

File: frame/randomness-collective-flip/src/lib.rs [hunk 2]
```diff
@@ -140,11 +162,13 @@ impl<T: Config> Randomness<T::Hash, T::BlockNumber> for Module<T> {
 mod tests {
 	use crate as pallet_randomness_collective_flip;
 	use super::*;
-	use sp_core::H256;
+
+    use sp_core::H256;
```
Comment (shawntabrizi) [inline frame/randomness-collective-flip/src/lib.rs:166]: some kind of spacing thing here

**Review by shawntabrizi — APPROVED** [review]
nits, but looks good otherwise

File: frame/randomness-collective-flip/src/lib.rs [hunk 3]
```diff
@@ -220,7 +220,7 @@ mod tests {
 		type OnSetCode = ();
 	}
 
-    impl pallet_randomness_collective_flip::Config for Test {} 
+    impl pallet_randomness_collective_flip::Config for Test {}
```
Comment (shawntabrizi) [inline frame/randomness-collective-flip/src/lib.rs:223]: Still a space problem here

Comment (koushiro) [reply frame/randomness-collective-flip/src/lib.rs]: It seems that there is a problem with my vim config, which has been fixed

File: frame/randomness-collective-flip/src/lib.rs [hunk 4]
```diff
@@ -85,21 +108,20 @@ decl_module! {
...
+        }
+    }
+
+	/// Series of block headers from the last 81 blocks that acts as random seed material. This
+	/// is arranged as a ring buffer with `block_number % 81` being the index into the `Vec` of
+	/// the oldest hash.
+    #[pallet::storage]
+    #[pallet::getter(fn random_material)]
+    pub(super) type RandomMaterial<T: Config> =
+        StorageValue<_, Vec<T::Hash>, ValueQuery>;
```
Comment (athei) [inline frame/randomness-collective-flip/src/lib.rs:121]: Some whitespace problems here.

**Review by athei — COMMENTED** [review]
The CI image will be fixed soon. I will restart the failing checks then.

File: frame/randomness-collective-flip/src/lib.rs (hunk 4 above)
Comment (koushiro) [reply frame/randomness-collective-flip/src/lib.rs]: Fixed, hope to use rustfmt in substrate ASAP :rofl:

**Review by athei — APPROVED** [review]
Thanks for your contribution.

**Review by thiolliere — APPROVED** [review]

**Comment by thiolliere** [comment]
bot merge

**Comment by JoshOrndorff** [comment]
@thiolliere I see you added the E1-runtimemigration tag here. I understand that as a chain using this pallet, I will need to do an onchain upgrade to run this new code. But there is no storage migration necessary right? I don't see any `on_runtime_upgrade` hooks.

**Comment by shawntabrizi** [comment]
@JoshOrndorff you may need to check that you imported this pallet with the same name as the `decl_storage` name.

This is the kind of migration you would need to look out for, and would be runtime specific

**Comment by thiolliere** [comment]
Yeah I updated the top message, the runtime migration is that the pallet prefix may have changed depending the construct_runtime declaration. If it has changed, then a migration must be made to move from the old prefix to the new prefix.

**Comment by JoshOrndorff** [comment]
Thanks for the help. I get it now. Let me make sure by making it totally concrete.

The old version of this pallet declared its storage prefix like this:
```rust
decl_storage! {
	trait Store for Module<T: Config> as RandomnessCollectiveFlip {
	  // --snip--
	}
}
```

So if my runtime declaration like this
```rust
// No migration necessary here because I happened to use the same `RandomnessCollectiveFlip` as in decl_storage
construct_runtime! {
	pub enum Runtime ... {
		System: frame_system::{Pallet, Call, Storage, Config, Event<T>},
		RandomnessCollectiveFlip: pallet_randomness_collective_flip::{Pallet, Call, Storage},
		// --snip--
	}
}

// Here a migration is necessary because I used a different name, `RandomnessBeacon`.
construct_runtime! {
	pub enum Runtime ... {
		System: frame_system::{Pallet, Call, Storage, Config, Event<T>},
		RandomnessBeacon: pallet_randomness_collective_flip::{Pallet, Call, Storage},
		// --snip--
	}
}
```

And if I were in the latter case, then I could write a runtime-level upgrade like this:
```rust
struct MigrateCollectiveFlip;
impl OnRuntimeUpgrade for MigrateCollectiveFlip {
  fn on_runtime_upgrade() -> Weight {
    frame_support::storage::migration::move_pallet(b"RandomnessCollectiveFlip", b"RandomnessBeacon");
  }
}
```

**Comment by thiolliere** [comment]
yes exactly

**Pull request merged**